src/compiler/checker.ts

```diff
@@ -24062,12 +24098,61 @@ namespace ts {
-                    || isThisTypeParameter(type)
-                    || type.flags & TypeFlags.Intersection && every((type as IntersectionType).types, t => t.symbol !== globalThisSymbol)) {
-                    return filterType(type, t => isTypePresencePossible(t, name, assumeTrue));
+            function widenTypeWithSymbol(type: Type, newSymbol: Symbol): Type {
+                // If type is this/any/unknown, it cannot be widened.
+                if ((type.flags & TypeFlags.AnyOrUnknown) || isThisTypeParameter(type)) {
+                    return type;
+                }
+                // If type is anonymous object, add the symbol directly
+                if (isObjectType(type) && type.objectFlags & ObjectFlags.Anonymous) {
+                    return widenObjectType(type, newSymbol);
+                }
+                // If type is intersection, add the symbol to the first anonymous object component of the intersection
+                if (isIntersectionType(type)) {
+                    const objectSubtype = type.types.find(t => getObjectFlags(t) & ObjectFlags.Anonymous) as ObjectType | undefined;
+                    if (objectSubtype) {
+                        const restOfIntersection = filterUnionOrIntersectionType(type, t => t !== objectSubtype);
+                        return createIntersectionType([restOfIntersection, widenObjectType(objectSubtype, newSymbol)]);
+                    }
+                }
+
+                // Otherwise, just add the new object type as an intersection
+                const newTypeWithSymbol = widenObjectType(createAnonymousType(undefined, createSymbolTable(), emptyArray, emptyArray, emptyArray), newSymbol);
+                return createIntersectionType([type, newTypeWithSymbol]);
+
+                function widenObjectType(type: ObjectType, newSymbol: Symbol): Type {
+                    const members = createSymbolTable();
+                    if (type.members !== undefined) {
+                        mergeSymbolTable(members, type.members);
+                    }
+                    members.set(newSymbol.escapedName, newSymbol);
+                    return createAnonymousType(undefined, members, type.callSignatures ?? emptyArray, type.constructSignatures ?? emptyArray, type.indexInfos ?? emptyArray);
+                }
+            }
+
+            function narrowOrWidenTypeByInKeyword(type: Type, name: __String, assumeTrue: boolean) {
+                // If type contains global this, don't touch it
+                if (type.symbol === globalThisSymbol
+                    || isUnionOrIntersectionType(type) && filterUnionOrIntersectionType(type, t => t.symbol === globalThisSymbol) !== neverType
+                   ) {
+                    return type;
+                }
+                const someDirectSubtypeContainsProp = getPropertyOfType(type, name, /* skipObjectFunctionPropertyAugment */ false, /* includePartialProperties */ true);

```

Comment:
Elsewhere where we've needed partial union properties, we just fetch them from the union (or intersection)'s property list directly, and have thus far avoided adding a parameter like this. But rather than checking for a matching concrete prop, I think pattern template literal index signatures will throw a wrench in this strategy - I think instead, you'll want to construct the single element `{prop: unknown}` type and see if it's comparable with the `type` here. I say comparable, because comparability should decompose the top-level unions and return true if any member matches. (Is there a reason to not use `isTypePresencePossible` directly?)

Reply:
Thanks, used `isTypePresencePossible` to get by without the additional parameter ;)